File: go/binary-search-tree/binary_search_tree.go
package binarysearchtree

type BinarySearchTree struct {
	left  *BinarySearchTree
	data  int
	right *BinarySearchTree
}

func NewBst(i int) *BinarySearchTree {
	return &BinarySearchTree{data: i}
}

func (bst *BinarySearchTree) Insert(i int) {
	var newTree **BinarySearchTree
	if i > bst.data {
		newTree = &bst.right
	} else {
		newTree = &bst.left
	}
	if *newTree == nil {
		*newTree = NewBst(i)
	} else {
		(*newTree).Insert(i)
	}
}

func (bst *BinarySearchTree) SortedData() []int {
	result := []int{}
	if bst == nil {
		return result
	}
	result = append(result, bst.left.SortedData()...)
	result = append(result, bst.data)
	result = append(result, bst.right.SortedData()...)
	return result
}

File: javascript/binary-search-tree/binary-search-tree.js
export class BinarySearchTree {
  constructor(data) {
    this._data = data;
  }

  get data() {
    return this._data;
  }
  
  get right() {
    return this._right;
  }

  get left() {
    return this._left;
  }

  insert(data) {
    let current = this;
    while (true) {
      if (data <= current._data) {
        if (!current._left) {
          current._left = new BinarySearchTree(data);
          break;
        } else {
          current = current._left;
        };
      } else {
        if (!current._right) {
          current._right = new BinarySearchTree(data);
          break;
        } else {
          current = current._right;
        };
      };
    };
  }

  each(func) {
    if (this._left) {
      this._left.each(func);
    };
    func(this._data);
    if (this._right) {
      this._right.each(func);
    };
  }
}
